mqtt-streaming/src/main/scala/akka/stream/alpakka/mqtt/streaming/impl/HighLevelMqttSource.scala

```diff
@@ -0,0 +1,158 @@
+  ): Source[Out, Future[immutable.Seq[(String, ControlPacketFlags)]]] = {
+    val createAckHandle: SourceQueueWithComplete[Command[Nothing]] => PartialFunction[Event[Nothing], Out] =
+      commands => {
+        case Event(publish @ Publish(_, _, Some(packetId), _), _) =>
+          createOut(publish,
+                    () =>
+                      commands
+                        .offer(Command(PubAck(packetId)))
+                        .map(_ => Done)(ExecutionContexts.sameThreadExecutionContext))
+        case Event(publish: Publish, _) =>
+          throw new RuntimeException(s"Received Publish without packetId in at-least-once mode: $publish")
+      }
+
+    createOnMaterialization[Out](mqttClientSession,
+                                 transportSettings,
+                                 restartSettings,
+                                 connectionSettings,
+                                 subscriptions,
+                                 createAckHandle)
+  }
+
+  private def createOnMaterialization[Out](
+      mqttClientSession: MqttClientSession,
+      transportSettings: MqttTransportSettings,
+      restartSettings: MqttRestartSettings,
+      connectionSettings: MqttConnectionSettings,
+      subscriptions: MqttSubscribe,
+      createAckHandle: SourceQueueWithComplete[
+        Command[Nothing]
+      ] => PartialFunction[Event[Nothing], Out]
+  ) =
+    Setup
+      .source { implicit materializer => implicit attributes =>
+        implicit val system: ActorSystem = materializer.system
+
+        constructInternals[Out](mqttClientSession,
+                                transportSettings,
+                                restartSettings,
+                                connectionSettings,
+                                subscriptions,
+                                createAckHandle)
+      }
+      .mapMaterializedValue(_.flatMap(identity)(ExecutionContexts.sameThreadExecutionContext))
+
+  private def constructInternals[Out](
+      mqttClientSession: MqttClientSession,
+      transport: MqttTransportSettings,
+      restartSettings: MqttRestartSettings,
+      connect: MqttConnectPacket,
+      subscribe: MqttSubscribe,
+      acknowledgeAndOut: SourceQueueWithComplete[Command[Nothing]] => PartialFunction[Event[Nothing], Out]
+  )(implicit system: ActorSystem, materializer: Materializer) = {
+    val mqttFlow: Flow[Command[Nothing], Either[MqttCodec.DecodeError, Event[Nothing]], NotUsed] = {
+      import restartSettings._
+      RestartFlow.onFailuresWithBackoff(minBackoff, maxBackoff, randomFactor, maxRestarts) { () =>
+        Mqtt
+          .clientSessionFlow(mqttClientSession)
+          .join(transport.connectionFlow())
+      }
+    }
+    val subscribed = Promise[immutable.Seq[(String, ControlPacketFlags)]]()
+
+    val subscribePacket = subscribe.controlPacket
+    val initCommands = immutable.Seq(
+      Command(connect.controlPacket),
+      Command(subscribePacket)
+    )
```

Comment:
Perhaps these should be prepended to the factory flow (L113-L115) so that they're sent on every connect? e.g. 

```scala
    val mqttFlow: Flow[Command[Nothing], Either[MqttCodec.DecodeError, Event[Nothing]], NotUsed] = {
      import restartSettings._
      RestartFlow.onFailuresWithBackoff(minBackoff, maxBackoff, randomFactor, maxRestarts) { () =>
        Flow[Command[Nothing]]
          .prepend(
            Source(
              immutable.Seq(
                Command(connect.controlPacket),
                Command(subscribePacket)
               )
            )
          )
          .via(
            Mqtt
              .clientSessionFlow(mqttClientSession)
              .join(transport.connectionFlow())
           )
      }
    }
```

Reply:
Yes, thank you.